IH-689: Autogenerate record_util.ml (where it's reasonable)

last-genius opened the pull request:
Best reviewed by commit - commit messages include the necessary context.

Changes can mostly be grouped into:
* Implementing autogeneration
* Removing manually-specified conversion functions that are identical to the automatically-generated ones
* Removing converters identical in function but not in form **(requires renaming usage sites)** - these are commit-per-function for ease of review, but can be squashed before merging
* Removing conversion functions that were not covered with tests **(requires also adding corresponding tests)** - these are commit-per-function
* Justifying the grim shadowing existence of the rest.

It might seem like a big patch, but most of it is removing and renaming things, with only a tiny bit of new code. Since Edwin's PR some time ago (#5349), we can now confirm the behaviour of the autogenerated code is the same as the old code, too!

It's better to autogenerate `record_util.ml` because we already have a lot of boilerplate that needs to be touched to add new things to the datamodel and CLI, this reduces the boring parts. In addition, the manual nature of this file caused duplicates (that existed since before git! And nobody noticed!), inconsistent behaviour (dots/dashes/underscores/case-sensitivity), odd cases (functions converting variants **not** in the datamodel), etc. The sooner we can get rid of the large part of this the better.

psafont commented on ocaml/idl/ocaml_backend/gen_api.ml:
```diff
@@ -354,6 +354,38 @@ let toposort_types highapi types =
   assert (List.sort compare result = List.sort compare types) ;
   result
 
+let gen_utils highapi =
```
Utils is not very useful name, maybe `gen_record_deser`?

psafont commented:
I really like the approach used here to keep maintenance of polymorphic variants in the datamodel to a minimum.

About the VM powerstate I'll wait for @robhoes to reply as he will know more about the history that needed the two additional states

minglumlu commented on ocaml/xapi/xapi_vm_clone.ml:
```diff
@@ -231,6 +231,8 @@ let quiesced = "quiesced"
 let snapshot_info ~power_state ~is_a_snapshot =
   let power_state_info =
+    (* Uses a shadowed function from record_util because xapi_vm_clone.mli
+       interface includes non-datamodel enum variants *)
```
I can't find where the non-datamodel enum variants are used. Are they really useful?

psafont commented:
Just in time for their 18th birthday...

last-genius commented:
Addressed the review comments. XenRT tests failed all over the place though, will re-run and try to find some time this week to look into them. Converting to draft for now.

robhoes commented:
A good simplification of the code, removing lots of lines. I can't see anything wrong with this.

last-genius commented:
@robhoes pointed out that XenRT, beside our own tests, rely on the string output of vm_power_state conversions to be lowercase - this is likely the cause for suite failures. I've removed two deprecated power states, and kept the lowercase function (and its users) as is.

last-genius commented on ocaml/xapi-cli-server/record_util.ml:
```diff
@@ -801,6 +786,8 @@ let string_to_vbd_type s =
 
+(* Intentional shadowing - some usage sites rely on the
+   non-capitalized string output of the conversion function. *)
 let power_to_string h =
```
Made this change.
